guide/adopting-the-configuration-cache.md

```diff
@@ -0,0 +1,384 @@
+
+The configuration phase typically runs faster than execution since heavy work belongs in the latter. However, without the cache, configuration phase is rerun with every single Gradle run. If you run a unit test multiple times, configuration is repeated, reproducing the same task graph every time. When the task itself is light, configuration can take up a large fraction of the latency. Configuration caching solves this by storing and reusing configuration results between runs, eliminating redundant work and speeding up iteration cycles.
+
+
+
+## Finding Configuration Cache problems
+
+To find Configuration Cache problems:
+
+1. Run `./gradlew build --configuration-cache`
+2. Check the output for any problems. If none appear, great! Your build is compatible.
+3. If not, fix the problems.
+4. Run `./gradlew build --configuration-cache` again. In our experience, this reveals additional issues not seen in the first run.
+
+The problems typically fall into three categories:
+
+1. **external process started `/usr/bin/git --version`**
+2. **cannot serialize object of type `org.gradle.api.internal.project.DefaultProject`, a subtype of `org.gradle.api.Project`, as these are not supported with the configuration cache**
+3. **invocation of `Task.project` at execution time is unsupported**
+
+The first error occurs when you start external processes during the configuration phase. Generally, it is preferred to run external processes in tasks with properly declared inputs and outputs to avoid unnecessary work when the task is UP-TO-DATE. 
+
+The second and third errors occur because Gradle tasks shouldn't take mutable types as input — `Gradle`, `Settings`, `Project`, `SourceSet`, or `Configuration`. Having these as inputs limits task concurrency (what if two tasks mutate a `Project` concurrently?) and prevents Gradle from serializing task inputs into the cache (they are too large to be serialized). 
+
+To solve the second and third errors, you can
+- Declare the smallest "surface area" your task needs as task input, e.g. take in `Property<String>` for project version, instead of taking in `Project` and accessing `project.version`
+- [Inject a service](https://docs.gradle.org/current/userguide/service_injection.html) which provides the information/operation you need
+
+
+## Solving Configuration Cache problems
+
+Let’s walk through some examples of fixing Configuration Cache problems
+
+> [!TIP]
+> If you have a huge Gradle projects with many tasks, you can adopt the Configuration Cache incrementally. First, resolve all configuration phase issues, as incremental adoption only works for the execution phase. Then, apply the [gradle-incremental-configuration-cache](https://github.com/palantir/gradle-incremental-configuration-cache) plugin. Gradually add tasks to the allow list as they become compatible.
```

Comment:
This seems somewhat out of place:

> First, resolve all configuration phase issues

This is what this whole section is about? Also have we discussed what the difference between configuration phase cc issues and execution phase cc issues are? There's lots to unpack in this tip that is not obvious to me.

This tip probably deserves its own section, at the bottom

Reply:
> This is what this whole section is about

Hmm, this section talks about both configuration-phase (e.g. external process started) and execution-phase (e.g. calling Task#project in @TaskAction) issues. 

> have we discussed what the difference between configuration phase cc issues and execution phase cc issues are?

I've added a clarification that this refers to problems like `external process started`.

Also, this guide is intended to be standalone, independent of `gradle-incremental-configuration-cache`. I don't think having a separate section for the plugin respects this.